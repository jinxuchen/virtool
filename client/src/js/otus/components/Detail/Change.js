import { get } from "lodash-es";
import React from "react";
import styled from "styled-components";

import { device, Icon, Label, ListGroupItem, RelativeTime } from "../../../base";

const methodIconProps = {
    add_isolate: {
        name: "flask",
        bsStyle: "primary"
    },
    create: {
        name: "plus-square",
        bsStyle: "primary"
    },
    create_sequence: {
        name: "dna",
        bsStyle: "primary"
    },
    edit: {
        name: "pencil-alt",
        bsStyle: "warning"
    },
    edit_isolate: {
        name: "flask",
        bsStyle: "warning"
    },
    edit_sequence: {
        name: "dna",
        bsStyle: "warning"
    },
    clone: {
        name: "clone",
        bsStyle: "primary"
    },
    import: {
        name: "file-import",
        bsStyle: "primary"
    },
    remote: {
        name: "link",
        bsStyle: "primary"
    },
    remove: {
        name: "trash",
        bsStyle: "danger"
    },
    remove_isolate: {
        name: "flask",
        bsStyle: "danger"
    },
    remove_sequence: {
        name: "dna",
        bsStyle: "danger"
    },
    set_as_default: {
        name: "star",
        bsStyle: "warning"
    },
    update: {
        name: "arrow-alt-circle-up",
        bsStyle: "warning"
    }
};

const getMethodIcon = ({ method_name }) => {
    const props = get(methodIconProps, method_name, {
        name: "exclamation-triangle",
        bsStyle: "danger"
    });

    return <Icon {...props} />;
};

const StyledChange = styled(ListGroupItem)`
    display: flex;
    flex-flow: row wrap;
    @media (max-width: ${device.tabelet}) {
        align-items: flex-start;
        display: flex;
        flex-flow: row wrap;
    }
`;

const IconDescriptionNameDate = styled.div`
    align-items: start;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    margin-left: 9px;
    width: 90%;

    @media (max-width: ${device.tabelet}) {
        flex-flow: row wrap;
        flex-direction: column;
    }
`;

const Name = styled.div`
    align-items: center;
    display: flex;
    flex-flow: row wrap;
    margin-right: 10px;
`;

const ChangeIcon = styled.div`
    display: flex;
    margin-right: 4px;
`;

const DateUserId = styled.div`
    @media (max-width: ${device.tabelet}) {
        font-size: 12px;
    }
`;

export class Change extends React.Component {
    handleRevert = () => {
        this.props.revert(this.props.otu.id, this.props.otu.version, this.props._id);
    };

    render() {
        return (
            <StyledChange>
                <Label>{this.props.otu.version}</Label>

                <IconDescriptionNameDate>
                    <Name>
                        <ChangeIcon>{getMethodIcon(this.props)}</ChangeIcon>
                        <span className="test-1">{this.props.description || "No Description"}</span>
                    </Name>
                    <DateUserId>
                        <RelativeTime style={{ fontSize: 12 }} time={this.props.created_at} /> by
                        <span> {this.props.user.id}</span>
                    </DateUserId>
                </IconDescriptionNameDate>
            </StyledChange>
        );
    }
}
