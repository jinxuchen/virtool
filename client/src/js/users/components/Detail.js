/**
 * @license
 * The MIT License (MIT)
 * Copyright 2015 Government of Canada
 *
 * @author
 * Ian Boyes
 *
 * @exports Users
 */

import { push } from "connected-react-router";
import { capitalize, get, map } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import {
    Flex,
    FlexItem,
    Icon,
    Identicon,
    InputError,
    LoadingPlaceholder,
    RemoveBanner,
    WarningAlert
} from "../../base";
import { listGroups } from "../../groups/actions";

import { editUser, getUser, removeUser } from "../actions";
import UserGroups from "./Groups";
import Password from "./Password";
import UserPermissions from "./Permissions";
import PrimaryGroup from "./PrimaryGroup";

export class UserDetail extends React.Component {
    componentDidMount() {
        this.props.onGetUser(this.props.match.params.userId);

        if (!this.props.groupsFetched) {
            this.props.onListGroups();
        }
    }

    handleSetPrimaryGroup = e => {
        const value = e.target.value === "none" ? "" : e.target.value;
        this.props.onSetPrimaryGroup(this.props.detail.id, value);
    };

    toggleAdmin = e => {
        this.props.onSetUserRole(this.props.detail.id, e.target.value === "Administrator");
    };

    handleRemove = () => {
        this.props.onRemoveUser(this.props.detail.id);
    };

    render() {
        if (this.props.error.length) {
            return (
                <WarningAlert level>
                    <Icon name="exclamation-circle" />
                    <span>
                        <strong>You do not have permission to manage users.</strong>
                        <span> Contact an administrator.</span>
                    </span>
                </WarningAlert>
            );
        }

        if (this.props.detail === null) {
            return <LoadingPlaceholder />;
        }

        const currentRole = this.props.detail.administrator ? "Administrator" : "Limited";

        const canModifyUser = this.props.activeUser !== this.props.detail.id && this.props.activeUserIsAdmin;

        return (
            <div>
                <Flex justifyContent="space-between">
                    <Flex alignItems="center">
                        <Identicon size={64} hash={this.props.detail.identicon} />
                        <FlexItem pad={10}>
                            <h5>
                                <strong>{this.props.detail.id}</strong>
                            </h5>
                        </FlexItem>
                        <FlexItem pad={10}>
                            {this.props.detail.administrator ? <Icon name="user-shield" bsStyle="primary" /> : null}
                        </FlexItem>
                    </Flex>
                    <Flex alignItems="center">
                        <FlexItem>
                            <Link to="/administration/users">Back To List</Link>
                        </FlexItem>
                    </Flex>
                </Flex>

                <div style={{ marginTop: "20px" }}>
                    <Password />

                    <UserGroups />

                    <PrimaryGroup />

                    <Flex alignItems="center" justifyContent="space-between">
                        <label>Permissions</label>
                        <small className="text-muted">Change group membership to modify permissions</small>
                    </Flex>
                    <UserPermissions permissions={this.props.detail.permissions} />

                    {canModifyUser ? (
                        <React.Fragment>
                            <label>User Role</label>
                            <InputError type="select" value={currentRole} onChange={this.toggleAdmin}>
                                <option key="admin" value="Administrator">
                                    Administrator
                                </option>
                                <option key="limit" value="Limited">
                                    Limited
                                </option>
                            </InputError>
                        </React.Fragment>
                    ) : null}

                    {canModifyUser ? (
                        <RemoveBanner
                            message="Click the Delete button to permanently remove this user."
                            buttonText="Delete"
                            onClick={this.handleRemove}
                        />
                    ) : null}
                </div>
            </div>
        );
    }
}

export const mapStateToProps = state => ({
    detail: state.users.detail,
    activeUser: state.account.id,
    activeUserIsAdmin: state.account.administrator,
    groups: state.groups.list,
    groupsFetched: state.groups.fetched,
    error: get(state, "errors.GET_USER_ERROR.message", "")
});

export const mapDispatchToProps = dispatch => ({
    onGetUser: userId => {
        dispatch(getUser(userId));
    },

    onRemoveUser: userId => {
        dispatch(removeUser(userId));
    },

    onClose: () => {
        dispatch(push("/administration/users"));
    },

    onSetPrimaryGroup: (userId, groupId) => {
        dispatch(editUser(userId, { primary_group: groupId }));
    },

    onSetUserRole: (userId, isAdmin) => {
        dispatch(editUser(userId, { administrator: isAdmin }));
    },

    onListGroups: () => {
        dispatch(listGroups());
    }
});

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(UserDetail);
