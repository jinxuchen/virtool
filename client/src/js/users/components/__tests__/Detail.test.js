import { UserDetail, mapStateToProps, mapDispatchToProps } from "../Detail";

describe("<UserDetail />", () => {
    let props;

    beforeEach(() => {
        props = {
            onGetUser: jest.fn(),
            match: {
                params: {
                    userId: "foo"
                }
            },
            groupsFetched: true,
            onListGroups: jest.fn(),
            onSetUserRole: jest.fn(),
            detail: {
                id: "foo",
                groups: "bar",
                administrator: "baz",
                identicon: "foo",
                primary_group: "Bar",
                permissions: { foo: "bar" }
            },
            onRemoveUser: jest.fn(),
            error: []
        };
    });

    it("should render", () => {
        const wrapper = shallow(<UserDetail {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("componentDidMount should call onGetUser", () => {
        expect(props.onGetUser).not.toHaveBeenCalled();
        shallow(<UserDetail {...props} />);
        expect(props.onGetUser).toHaveBeenCalledWith("foo");
    });

    it("componentDidMount should call onListGroups", () => {
        expect(props.onListGroups).not.toHaveBeenCalled();
    });
    it("componentDidMount should not call onListGroups when [props.groupsFetched = false]", () => {
        props.groupsFetched = false;
        shallow(<UserDetail {...props} />);
        expect(props.onListGroups).toHaveBeenCalled();
    });
});

describe("mapStateToProps", () => {
    const state = {
        users: {
            detail: "foo"
        },
        account: {
            id: "foo",
            administrator: true
        },
        groups: {
            list: "foo",
            fetched: true
        }
    };
    const props = mapStateToProps(state);

    it("should renturn props", () => {
        expect(props).toEqual({
            activeUser: "foo",
            detail: "foo",
            activeUserIsAdmin: true,
            groupsFetched: true,
            error: ""
        });
    });
});

describe("mapDispatchToProps", () => {
    let dispatch;
    let result;

    beforeEach(() => {
        dispatch = jest.fn();
        result = mapDispatchToProps(dispatch);
    });

    it("should return onGetUser() in props", () => {
        const userId = "foo";
        result.onGetUser(userId);
        expect(dispatch).toHaveBeenCalledWith({
            type: "GET_USER_REQUESTED",
            userId
        });
    });
    it("should return onRemoveUser() in props", () => {
        const userId = "foo";
        result.onRemoveUser(userId);
        expect(dispatch).toHaveBeenCalledWith({
            type: "REMOVE_USER_REQUESTED",
            userId
        });
    });

    it("should return onSetPrimaryGroup() in props", () => {
        const userId = "foo";
        const groupId = "bar";
        result.onSetPrimaryGroup(userId, groupId);
        expect(dispatch).toHaveBeenCalledWith({
            type: "EDIT_USER_REQUESTED",
            update: { primary_group: "bar" },
            userId: "foo"
        });
    });
    it("should return onSetUserRole() in props", () => {
        const userId = "foo";
        const isAdmin = true;
        result.onSetUserRole(userId, isAdmin);
        expect(dispatch).toHaveBeenCalledWith({
            type: "EDIT_USER_REQUESTED",
            update: {
                administrator: true
            },
            userId: "foo"
        });
    });
    it("should return onListGroups() in props", () => {
        result.onListGroups();
        expect(dispatch).toHaveBeenCalledWith({
            type: "LIST_GROUPS_REQUESTED"
        });
    });
});
