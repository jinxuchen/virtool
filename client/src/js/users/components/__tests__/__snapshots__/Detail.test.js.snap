// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserDetail /> should render 1`] = `
<div>
  <Flex
    alignContent="stretch"
    alignItems="stretch"
    direction="row"
    justifyContent="space-between"
    wrap="nowrap"
  >
    <Flex
      alignContent="stretch"
      alignItems="center"
      direction="row"
      justifyContent="flex-start"
      wrap="nowrap"
    >
      <Identicon
        hash="foo"
        size={64}
      />
      <FlexItem
        alignSelf={null}
        basis="auto"
        grow={0}
        pad={10}
        shrink={1}
      >
        <h5>
          <strong>
            foo
          </strong>
        </h5>
      </FlexItem>
      <FlexItem
        alignSelf={null}
        basis="auto"
        grow={0}
        pad={10}
        shrink={1}
      >
        <Icon
          bsStyle="primary"
          faStyle="fas"
          fixedWidth={false}
          name="user-shield"
          pullRight={false}
        />
      </FlexItem>
    </Flex>
    <Flex
      alignContent="stretch"
      alignItems="center"
      direction="row"
      justifyContent="flex-start"
      wrap="nowrap"
    >
      <FlexItem
        alignSelf={null}
        basis="auto"
        grow={0}
        shrink={1}
      >
        <Link
          to="/administration/users"
        >
          Back To List
        </Link>
      </FlexItem>
    </Flex>
  </Flex>
  <div
    style={
      Object {
        "marginTop": "20px",
      }
    }
  >
    <Connect(Password) />
    <GroupPrimaryPermissions />
  </div>
</div>
`;
